### PR TITLE
Use "socket" as default comm in distributed lightgbm training + prepare for training framework comparison

### PR DESCRIPTION
**Pull request by jfomhover**
Our current LightGBM training relies on a custom container of LightGBM building mpi. This is not the purely default LightGBM installation. This PR removes this dependency and uses plain out-of-the-box lightgbm for distributed training (relying on "socket").

It also provides a configuration field `framework` in the lightgbm training config file to switch between socket and mpi builds. We're expecting to use this to compare different lightgbm implementations (socket, mpi, ray...) in our benchmark.

## Modifications

- modify `src/common/distributed.py` to introduce 2 "drivers", one for socket one for mpi, add an argument to switch between those two and provided required values (machines, listen port)
- modify lightgbm train to align with these drivers and the new multinode config format
- modify lightgbm_training.py pipeline to introduce "framework" field and switch between different implementations

### CODE REVIEW

File: src/scripts/training/lightgbm_python/train.py [hunk 1]
```diff
@@ -230,26 +249,26 @@ def run(self, args, logger, metrics_logger, unknown_args):
...
 
         logger.info(f"Training LightGBM with parameters: {lgbm_params}")
-        with metrics_logger.log_time_block("time_training", step=mpi_config.world_rank):
+        with metrics_logger.log_time_block("time_training", step=multinode_config.world_rank):
             booster = lightgbm.train(
                 lgbm_params,
                 train_data,
                 valid_sets = val_datasets,
-                callbacks=[callbacks_handler.callback]
+                #callbacks=[callbacks_handler.callback]
```
Comment (jfomhover) [inline src/scripts/training/lightgbm_python/train.py:257]: this is a typo, I should not have commented this line, fixing it asap

File: src/common/tasks.py [hunk 2]
```diff
@@ -132,6 +133,7 @@ class lightgbm_training_output_variant_parameters:
 @dataclass
 class training_variant:
     # three below are mandatory sections of the variant config
```
Comment (AliMahmoudzadeh) [inline src/common/tasks.py:135]: four?

**Review by AliMahmoudzadeh — APPROVED** [review]
looked good to me

File: src/common/tasks.py (hunk 2 above)
Comment (jfomhover) [reply src/common/tasks.py:135]: good catch! :-D

**Pull request merged**